ops/model.py

```diff
@@ -663,3 +664,40 @@ def network_get(self, endpoint_name, relation_id=None):
             if 'relation not found' in str(e):
                 raise RelationNotFoundError() from e
             raise
+
+    def add_metrics(self, metrics, labels=None):
+        cmd = ['add-metric']
+        key_re = re.compile(r'^[a-zA-Z][a-zA-Z0-9_]*$')
+
+        def is_valid_metric_value(value):
+            try:
+                float(value)
+                return True
+            except ValueError:
+                return False
+
+        def is_valid_label_value(value):
+            # Label values cannot be empty, contain commas or equal signs as those are used by add-metric as separators.
+            return value and all(c not in str(value) for c in ',=')
+
+        if labels:
+            label_args = []
+            for k, v in labels.items():
+                if not key_re.match(k):
+                    raise ModelError(f'invalid label key "{k}": must start from an ASCII letter and contain alphanumeric characters or underscores only')
+                elif not is_valid_label_value(v):
+                    raise ModelError('metric label values must not contain "," or "=".')
```

Comment:
No dots at the end of error messages.

Reply:
https://github.com/canonical/operator/pull/140/commits/e67e3c7534ed439b2208a261714a5bb88fe4ee15